packages/material-ui-styled-engine/src/index.js

```diff
@@ -1 +1,2 @@
 export { default } from '@emotion/styled';
+export * from 'emotion-theming';
```

Comment:
We should specify what we need in case `emotion-theming` adds new modules that we don't need.

Reply:
What is `emotion-theming` for?

Reply:
By adding the emotion's/styled-components' `ThemeProvider` in our `ThemeProvider` component, developers no longer need to specify `ThemeProvider` from emotion/styled component together with ours. Also, in the `styled` utility when custom theme is used, it will automatically be available on `props.theme`. In addition to this, when no `ThemeProvider` is used, the `styled` utility uses the `defaultTheme`

Reply:
What if they already use `emotion-theming`?

Reply:
I think that using a different ThemeProvider (enforcing the usage of two) makes incremental adoption easier. Developers don't have to worry about conflicts of theme structure.

However, for using the css prop, it seems required https://emotion.sh/docs/theming#css-prop. 

It seems that theme-ui imports the ThemeContext of emotion directly (without emotion-theming), https://github.com/system-ui/theme-ui/blob/af57d4ce97c0e8a3289115f33cef1d17bc65adad/packages/core/src/index.ts#L3.

Reply:
Just so we are all on the same page as to what problem I am trying to solve. Here is a code snippet of how developers would define their custom components with the changes on the Provider:

```jsx
import styled from '@material-ui/core/experimentalStyled';
import { createMuiTheme } from '@material-ui/core/styles';
import { ThemeProvider } from '@material-ui/styles';
// in case of styled-components they would import the ThemeProvider from styled-components
import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming';

// how the overrides should be resolved
const testOverridesResolver = (props, styles) => ({
  ...styles.root,
  ...(props.variant && styles[props.variant]),
});

// how the component is defined
const Test = styled(
  'div',
  { shouldForwardProp: (prop) => prop !== 'variant' && prop !== 'size' },
  { muiName: 'MuiTest', overridesResolver: testOverridesResolver },
)`
  width: 200px;
  height: 300px;
`;

// some theme overrides
const theme = createMuiTheme({
      components: {
        MuiTest: {
          styleOverrides: {
            root: {
              width: '250px',
            },
            rect: {
              height: '250px',
            },
          },
        },
      },
    });

// now for the overrides to actually work without using the ThemeProvider from emotion, they will need to do this
const Example = props => (
  <ThemeProvider theme={theme}>
    <EmotionThemeProvider theme={theme}>
      <Test>Test</Test>
    </EmotionThemeProvider>
  </ThemeProvider>
)
```

If we do the changes that are currently on the PR, they wouldn't need to use the emotion/styled-components ThemeProvider. More over, without these changes, our components would respect the theme from the ThemeProvider we expose, and their custom components (for example created on top of div), would respect the emotion/styled-components ThemeProvider theme.

So there is a simple fix - we can document always when you use our ThemeProvider, use the ThemeProvider from emotion/styled-components. If we are ok with this and document it properly, none of the changes from this PR will be required expect for the `defaultTheme`, which anyway is a change that we want :)

Reply:
Is it accurate to say that developers could write their whole application with `muiStyled` and never have to use emotion's theme provider nor styled-components' theme provider?

Reply:
If we do these changes yes, that's what I was trying to solve. If we don't they would need to use both provider (which again if we document shouldn't be a problem) I am just trying to emphasize on what problem I am solving.

Reply:
> If we don't they would need to use both provider

I don't understand this part. Why is only using mui's styled and mui's theme provider not working?

Reply:
Because `muiStyled` is a simple wrapper around emotions/styled-components styled utility. The component generated there is subscribed to their own theme context, not ours. 

The simplest way of fixing this is for us to always include their Provider when we are rendering ours. If we do not do this, the only way I see this to be solved is by using a wrapper component on the component they are generating and consuming there our theme context - have to test this in order to be sure that it works.

We cannot simply use `useTheme` in muiStyled as it is not a component, we need to create Wrapper for it, which will mean two components instead of one for each styled component.

Or maybe I am missing some way of how we can differently fix this

Reply:
If we go the other way around, we could ask people to always use emotion/styled-components ThemeProvider and we can just substribe to their context instead of our own theme context.

Reply:
Just a clarification, for our components, like `Slider`, `Button`, `muiStyled` + our `ThemeProvider` is working, it won't work for the scenarios of using `muiStyled` + `div`, `span` or other primitive elements as we do not have access to change the behavior there.

Reply:
Oh, I see, thanks for the clarification. +1 for including the styled-engine theme provider in Material-UI's ThemeProvider component. It's simpler. For the theme structure conflict, developers can workaround the issue by doing MuiThemeProvider > Emotion ThemeProvider.

The only point I would suggest to explore is: Using ThemeContext exported from `@emotion/core` vs. `emotion-theming`.

Reply:
> The only point I would suggest to explore is: Using the ThemeContext exported from @emotion/core vs. emotion-theming.

Yeah, let me spend some time on this, just to make sure I am not missing something 👍

Reply:
Refactored to use only the ThemeContext (both emotion & styled-components expose this). Tested with both engines, tests are passing, so I think everything is covered. No new dependencies added with this 👍

Reply:
> No new dependencies added with this 

Nice :).